Export directly into ZIP archive: Import fails for attachments with whitespace in file name

## How to reproduce

1. Enonic XP 7.16.4 running as docker container (not sandbox - see note at bottom of this issue description)
2. Data Toolbox 5.2.1
3. Create a file "file name.txt" and upload it to Content Studio. What's in the file and its file type is irrelevant, just that it has one or more spaces in the file name and that it will be stored as an attachment in Content Studio
4. Export this file with Data Toolbox, making sure that "Archive" is CHECKED
5. In a separate environment with same type of hosting, XP 7.16.4, and Data Toolbox 5.2.1, upload export and try to import

Import fails with the following stack trace:
```
2026-07-03 09:02:36,150 ERROR s.r.enonic.datatoolbox.RcdScriptBean - Error while loading export
java.lang.IllegalArgumentException: Illegal character in opaque part at index 90: jar:file:///enonic-xp/home/data/export/file%20name.txt-draft-2026-07-03T10-58-40.zip!/file name.txt-draft-2026-07-03T10-58-40/file name.txt
	at java.base/java.net.URI.create(URI.java:932)
	at com.enonic.xp.core.impl.export.reader.ZipVirtualFile.getUrl(ZipVirtualFile.java:131)
	at com.enonic.xp.core.impl.export.reader.ExportReader.lambda$getChildren$0(ExportReader.java:33)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at com.enonic.xp.core.impl.export.NodeImporter.importFromDirectoryLayout(NodeImporter.java:121)
	at com.enonic.xp.core.impl.export.NodeImporter.execute(NodeImporter.java:102)
	at com.enonic.xp.core.impl.export.ExportServiceImpl.importNodes(ExportServiceImpl.java:149)
	at systems.rcd.enonic.datatoolbox.RcdExportScriptBean.load(RcdExportScriptBean.java:334)
	at systems.rcd.enonic.datatoolbox.RcdExportScriptBean.lambda$load$1(RcdExportScriptBean.java:260)
	at com.enonic.xp.context.ContextImpl.runWith(ContextImpl.java:84)
	at systems.rcd.enonic.datatoolbox.RcdExportScriptBean.lambda$load$0(RcdExportScriptBean.java:255)
	at systems.rcd.enonic.datatoolbox.RcdScriptBean.runSafely(RcdScriptBean.java:46)
	at systems.rcd.enonic.datatoolbox.RcdScriptBean.runSafely(RcdScriptBean.java:39)
	at systems.rcd.enonic.datatoolbox.RcdExportScriptBean.load(RcdExportScriptBean.java:252)
	at org.openjdk.nashorn.scripts/org.openjdk.nashorn.internal.scripts.Script$Recompilation$4602$503A$node_import.L:1#post#task(systems.rcd.enonic.datatoolbox:/services/node-import/node-import.js:15)
	at org.openjdk.nashorn.javaadapters.java_util_function_Function.apply(Unknown Source)
	at com.enonic.xp.lib.task.TaskWrapper.runTask(TaskWrapper.java:38)
	at com.enonic.xp.task.TaskProgressReporterContext.lambda$withContext$0(TaskProgressReporterContext.java:36)
	at com.enonic.xp.lib.task.TaskWrapper.run(TaskWrapper.java:31)
	at com.enonic.xp.impl.task.DescribedTaskImpl.run(DescribedTaskImpl.java:56)
	at com.enonic.xp.impl.task.TaskRunnable.lambda$doRun$2(TaskRunnable.java:57)
	at com.enonic.xp.context.ContextImpl.runWith(ContextImpl.java:84)
	at com.enonic.xp.impl.task.TaskRunnable.doRun(TaskRunnable.java:57)
	at com.enonic.xp.trace.Tracer.trace(Tracer.java:128)
	at com.enonic.xp.impl.task.TaskRunnable.run(TaskRunnable.java:39)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.URISyntaxException: Illegal character in opaque part at index 90: jar:file:///enonic-xp/home/data/export/file%20name.txt-draft-2026-07-03T10-58-40.zip!/file name.txt-draft-2026-07-03T10-58-40/file name.txt
	at java.base/java.net.URI$Parser.fail(URI.java:2995)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3166)
	at java.base/java.net.URI$Parser.parse(URI.java:3202)
	at java.base/java.net.URI.<init>(URI.java:645)
	at java.base/java.net.URI.create(URI.java:930)
	... 35 common frames omitted
```

## If part of a larger export with children
If the import is part of a larger export of multiple content where some of them have attachments, Data Toolbox will complete the import but with many errors such as this:
```
Errors: [
  "Error when parsing children of aldersvennlig-master-archive-2026-07-01T07-55-31/aldersvennlig/nyheter/eldreradene-etterlyser-tidligere-involvering-og-tettere-samarbeid-med-kommunen - java.lang.IllegalArgumentException: Illegal character in opaque part at index 247: jar:file:///enonic-xp/home/data/export/aldersvennlig-master-archive-2026-07-01T07-55-31.zip!/aldersvennlig-master-archive-2026-07-01T07-55-31/aldersvennlig/nyheter/eldreradene-etterlyser-tidligere-involvering-og-tettere-samarbeid-med-kommunen/Kurs for eldrerådsledere.jpg",
  "Error when parsing children of aldersvennlig-master-archive-2026-07-01T07-55-31/aldersvennlig/nyheter/ny-undersokelse-en-av-fire-i-media-er-over-60-ar - java.lang.IllegalArgumentException: Illegal character in opaque part at index 218: jar:file:///enonic-xp/home/data/export/aldersvennlig-master-archive-2026-07-01T07-55-31.zip!/aldersvennlig-master-archive-2026-07-01T07-55-31/aldersvennlig/nyheter/ny-undersokelse-en-av-fire-i-media-er-over-60-ar/Eldre synlighet i media_Unsplash.jpg",
(…)
```

Note one difference here is that the single import contained URI-encoded whitespace `%20` followed by the file name which was _not_ URI-encoded, but in the large import shown directly above, there are no instances of URI-encoded whitespaces.

## Sandbox behavior
In Enonic Sandbox on MacOS, if you try to follow the "steps to reproduce" from Sandbox A => Sandbox B, when you try to upload the export into Sandbox B in Data Toolbox, this error is thrown:
```
2026-07-03 13:42:51,122 ERROR s.r.enonic.datatoolbox.RcdScriptBean - Error while unarchiving exports
java.lang.RuntimeException: java.nio.file.NoSuchFileException: /Users/bhj/.enonic/distributions/enonic-xp-mac-arm64-sdk-7.16.4/home/work/export-archives-6386056718152004899/file name.txt-draft-2026-07-03T13-40-02.zip15927655351357408652.tmp -> /Users/bhj/.enonic/sandboxes/helsedir2/home/data/export/file name.txt-draft-2026-07-03T13-40-02.zip
	at systems.rcd.enonic.datatoolbox.RcdDataScriptBean.lambda$unarchive$0(RcdDataScriptBean.java:87)
	at systems.rcd.enonic.datatoolbox.RcdScriptBean.runSafely(RcdScriptBean.java:46)
	at systems.rcd.enonic.datatoolbox.RcdDataScriptBean.unarchive(RcdDataScriptBean.java:80)
	at org.openjdk.nashorn.scripts/org.openjdk.nashorn.internal.scripts.Script$Recompilation$601$570A$export_upload.L:1#post#task(systems.rcd.enonic.datatoolbox:/services/export-upload/export-upload.js:21)
	at org.openjdk.nashorn.javaadapters.java_util_function_Function.apply(Unknown Source)
	at com.enonic.xp.lib.task.TaskWrapper.runTask(TaskWrapper.java:38)
	at com.enonic.xp.task.TaskProgressReporterContext.lambda$withContext$0(TaskProgressReporterContext.java:36)
	at com.enonic.xp.lib.task.TaskWrapper.run(TaskWrapper.java:31)
	at com.enonic.xp.impl.task.DescribedTaskImpl.run(DescribedTaskImpl.java:56)
	at com.enonic.xp.impl.task.TaskRunnable.lambda$doRun$2(TaskRunnable.java:57)
	at com.enonic.xp.context.ContextImpl.runWith(ContextImpl.java:84)
	at com.enonic.xp.impl.task.TaskRunnable.doRun(TaskRunnable.java:57)
	at com.enonic.xp.trace.Tracer.trace(Tracer.java:128)
	at com.enonic.xp.impl.task.TaskRunnable.run(TaskRunnable.java:39)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.nio.file.NoSuchFileException: /Users/bhj/.enonic/distributions/enonic-xp-mac-arm64-sdk-7.16.4/home/work/export-archives-6386056718152004899/file name.txt-draft-2026-07-03T13-40-02.zip15927655351357408652.tmp -> /Users/bhj/.enonic/sandboxes/helsedir2/home/data/export/file name.txt-draft-2026-07-03T13-40-02.zip
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixFileSystem.move(UnixFileSystem.java:943)
	at java.base/sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:309)
	at java.base/java.nio.file.Files.move(Files.java:1430)
	at systems.rcd.enonic.datatoolbox.RcdDataScriptBean.lambda$unarchive$0(RcdDataScriptBean.java:85)
	... 16 common frames omitted
```

--- a/modules/core/core-export/src/main/java/com/enonic/xp/core/impl/export/reader/ExportReader.java
+++ b/modules/core/core-export/src/main/java/com/enonic/xp/core/impl/export/reader/ExportReader.java
@@ -1,6 +1,5 @@
 package com.enonic.xp.core.impl.export.reader;
 
-import java.net.URL;
 import java.util.stream.Stream;
 
 import com.enonic.xp.vfs.VirtualFile;
@@ -10,11 +9,6 @@ import static com.enonic.xp.core.impl.export.ExportConstants.SYSTEM_FOLDER_NAME;
 
 public class ExportReader
 {
-    private static String getLastBitFromUrl( final URL url )
-    {
-        return url.toString().replaceFirst( ".*/([^/?]+).*", "$1" );
-    }
-
     public long getNodeFileCount( final VirtualFile virtualFile )
     {
         long nodeFileCount = virtualFile.getChildren().stream().mapToLong( this::getNodeFileCount ).sum();
@@ -29,7 +23,6 @@ public class ExportReader
     {
         return parent.getChildren()
             .stream()
-            .filter( VirtualFile::isFolder )
-            .filter( ( folder ) -> !getLastBitFromUrl( folder.getUrl() ).equals( SYSTEM_FOLDER_NAME ) );
+            .filter( VirtualFile::isFolder ).filter( ( folder ) -> !folder.getName().equals( SYSTEM_FOLDER_NAME ) );
     }
 }

--- a/modules/core/core-export/src/main/java/com/enonic/xp/core/impl/export/reader/ZipVirtualFile.java
+++ b/modules/core/core-export/src/main/java/com/enonic/xp/core/impl/export/reader/ZipVirtualFile.java
@@ -5,6 +5,7 @@ import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.net.MalformedURLException;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.charset.StandardCharsets;
@@ -128,11 +129,12 @@ public class ZipVirtualFile
     {
         try
         {
-            return URI.create( "jar:" + zipFilePath.toAbsolutePath().toUri() + "!/" + entryPath ).toURL();
+            return URI.create( "jar:" + zipFilePath.toAbsolutePath().toUri() + "!/" + new URI( null, null, entryPath, null ).getRawPath() )
+                .toURL();
         }
-        catch ( MalformedURLException e )
+        catch ( MalformedURLException | URISyntaxException e )
         {
-            throw new UncheckedIOException( e );
+            throw new UncheckedIOException( new IOException( e ) );
         }
     }
 

--- a/modules/core/core-export/src/test/java/com/enonic/xp/core/impl/export/reader/ExportReaderTest.java
+++ b/modules/core/core-export/src/test/java/com/enonic/xp/core/impl/export/reader/ExportReaderTest.java
@@ -1,0 +1,69 @@
+package com.enonic.xp.core.impl.export.reader;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import com.enonic.xp.core.impl.export.writer.ZipExportWriter;
+import com.enonic.xp.vfs.VirtualFile;
+import com.enonic.xp.vfs.VirtualFiles;
+
+import static java.nio.file.Files.createDirectories;
+import static java.nio.file.Files.writeString;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class ExportReaderTest
+{
+    @TempDir
+    Path tempDir;
+
+    @Test
+    void testGetChildrenWithWhitespaceInFolderNames()
+        throws IOException
+    {
+        final String exportName = "my-export";
+        final Path baseDir = tempDir.resolve( exportName );
+
+        try (ZipExportWriter writer = ZipExportWriter.create( tempDir, exportName ))
+        {
+            writer.writeElement( baseDir.resolve( "_" ).resolve( "node.xml" ), "<root/>" );
+            writer.writeElement( baseDir.resolve( "my content" ).resolve( "_" ).resolve( "node.xml" ), "<child/>" );
+            writer.writeElement( baseDir.resolve( "export.properties" ), "xp.version=1.0" );
+        }
+
+        final Path zipFile = tempDir.resolve( exportName + ".zip" );
+        final VirtualFile root = ZipVirtualFile.from( zipFile );
+
+        final ExportReader exportReader = new ExportReader();
+
+        final List<VirtualFile> children = assertDoesNotThrow( () -> exportReader.getChildren( root ).collect( Collectors.toList() ) );
+
+        // The "_" system folder is filtered out; the whitespace-named content folder is kept.
+        assertEquals( 1, children.size() );
+        assertTrue( children.stream().anyMatch( child -> "my content".equals( child.getName() ) ) );
+    }
+
+    @Test
+    void testGetChildrenFiltersSystemFolderForDirectoryExport()
+        throws IOException
+    {
+        final Path exportDir = tempDir.resolve( "my-export" );
+        createDirectories( exportDir.resolve( "_" ) );
+        writeString( exportDir.resolve( "_" ).resolve( "node.xml" ), "<root/>" );
+        createDirectories( exportDir.resolve( "mynode" ).resolve( "_" ) );
+        writeString( exportDir.resolve( "mynode" ).resolve( "_" ).resolve( "node.xml" ), "<child/>" );
+
+        final VirtualFile root = VirtualFiles.from( exportDir );
+
+        final List<VirtualFile> children = new ExportReader().getChildren( root ).collect( Collectors.toList() );
+
+        assertEquals( 1, children.size() );
+        assertTrue( children.stream().anyMatch( child -> "mynode".equals( child.getName() ) ) );
+    }
+}

--- a/modules/core/core-export/src/test/java/com/enonic/xp/core/impl/export/reader/ZipVirtualFileTest.java
+++ b/modules/core/core-export/src/test/java/com/enonic/xp/core/impl/export/reader/ZipVirtualFileTest.java
@@ -16,7 +16,9 @@ import com.enonic.xp.vfs.VirtualFile;
 import com.enonic.xp.vfs.VirtualFilePath;
 import com.enonic.xp.vfs.VirtualFilePaths;
 
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -309,6 +311,86 @@ class ZipVirtualFileTest
         final VirtualFile file = current.resolve( VirtualFilePaths.from( "file.txt", "/" ) );
         assertTrue( file.exists() );
         assertEquals( "deep content", file.getCharSource().read() );
+    }
+
+    @Test
+    void testGetUrlEncodesWhitespaceInEntryAndArchivePath()
+        throws IOException
+    {
+        final Path subDir = tempDir.resolve( "my sub dir" );
+        final String exportName = "my-export";
+        final Path baseDir = subDir.resolve( exportName );
+
+        try (ZipExportWriter writer = ZipExportWriter.create( subDir, exportName ))
+        {
+            writer.writeElement( baseDir.resolve( "mynode" ).resolve( "_" ).resolve( "bin" ).resolve( "my photo.jpg" ), "binary" );
+            writer.writeElement( baseDir.resolve( "export.properties" ), "xp.version=1.0" );
+        }
+
+        final Path zipFile = subDir.resolve( exportName + ".zip" );
+        final VirtualFile root = ZipVirtualFile.from( zipFile );
+
+        final VirtualFile attachment = root.resolve( VirtualFilePaths.from( exportName + "/mynode/_/bin/my photo.jpg", "/" ) );
+        assertTrue( attachment.exists() );
+
+        final String url = assertDoesNotThrow( () -> attachment.getUrl().toString() );
+
+        // The whitespace in the entry name is encoded.
+        assertTrue( url.contains( "my%20photo.jpg" ), "entry whitespace should be encoded: " + url );
+        // The whitespace in the archive's own path stays single-encoded (not double-encoded to %2520).
+        assertTrue( url.contains( "my%20sub%20dir" ), "archive path whitespace should be encoded: " + url );
+        assertFalse( url.contains( "%2520" ), "archive path whitespace must not be double-encoded: " + url );
+    }
+
+    @Test
+    void testGetUrlEncodesReservedAndNonAsciiCharacters()
+        throws IOException
+    {
+        final String exportName = "my-export";
+        final Path baseDir = tempDir.resolve( exportName );
+        final String fileName = "50% off #1 résumé.pdf";
+
+        try (ZipExportWriter writer = ZipExportWriter.create( tempDir, exportName ))
+        {
+            writer.writeElement( baseDir.resolve( "mynode" ).resolve( "_" ).resolve( "bin" ).resolve( fileName ), "binary" );
+            writer.writeElement( baseDir.resolve( "export.properties" ), "xp.version=1.0" );
+        }
+
+        final Path zipFile = tempDir.resolve( exportName + ".zip" );
+        final VirtualFile root = ZipVirtualFile.from( zipFile );
+
+        final VirtualFile attachment = root.resolve( VirtualFilePaths.from( exportName + "/mynode/_/bin/" + fileName, "/" ) );
+        assertTrue( attachment.exists() );
+
+        final String url = assertDoesNotThrow( () -> attachment.getUrl().toString() );
+
+        assertTrue( url.contains( "%25" ), "'%' must be encoded: " + url );
+        assertTrue( url.contains( "%23" ), "'#' must be encoded: " + url );
+        assertTrue( url.contains( "%20" ), "whitespace must be encoded: " + url );
+        assertTrue( url.contains( "résumé" ), "non-ASCII is valid in a URI and should be preserved: " + url );
+        assertFalse( url.contains( "#" ), "no raw '#' should remain: " + url );
+        assertFalse( url.contains( " " ), "no raw whitespace should remain: " + url );
+    }
+
+    @Test
+    void testGetUrlForRootEntry()
+        throws IOException
+    {
+        final String exportName = "my-export";
+        final Path baseDir = tempDir.resolve( exportName );
+
+        try (ZipExportWriter writer = ZipExportWriter.create( tempDir, exportName ))
+        {
+            writer.writeElement( baseDir.resolve( "_" ).resolve( "node.xml" ), "<root/>" );
+            writer.writeElement( baseDir.resolve( "export.properties" ), "xp.version=1.0" );
+        }
+
+        final Path zipFile = tempDir.resolve( exportName + ".zip" );
+        final VirtualFile root = ZipVirtualFile.from( zipFile );
+
+        final String url = assertDoesNotThrow( () -> root.getUrl().toString() );
+        assertTrue( url.startsWith( "jar:" ), url );
+        assertTrue( url.contains( ".zip!/" + exportName ), url );
     }
 
     private void addZipEntry( final ZipOutputStream zos, final String path, final String content )

